Be/about pses

beccaelenzil opened the pull request:
Update About PSEs to reflect PSE01 rework: https://github.com/Ada-Developers-Academy/core-pses/pull/18

Added Example PSEs with Rock, Paper, Scissors example:
- High level: What do we think of this example? Rock, Paper, Scissors is a project thread in the ada build curriculum
- Question/idea, should we include all the example answers here as a resource, or make this a classroom activity and move the answers to an instructor doc?
- I had trouble coming up with clarifying questions. Ideas appreciated!

alope107 commented on about-pses/example-pse.md:
````diff
@@ -1,90 +1,170 @@
-    - Output: `{name: "Crow's Nest", rating: 1}`
+```
+1. How should the function handle invalid user input (i.e. player_1 = "lizards")?
+2. 
+3.
+4.
+5. 
````
Additional questions:
- Should a user automatically lose if they have invalid input?
- What should happen if both users have invalid input?
- Does capitalization matter?
- Does extra whitespace / punctuation matter?
- Should anything be printed to the console?
- Is the speed / memory usage of this function important?
- Should there be a default value for each of the arguments?

...but this question is so small and well-defined that coming up with this many clarifying questions _is_ really hard. Maybe for this one say that they only need to come up with one or two clarifying questions for this PSE but that later problems will be larger and they'll be expected to come up with more?

alope107 commented on about-pses/example-pse.md:
```diff
@@ -1,90 +1,170 @@
+        else:
+            return "Player 2 wins!"
+    # player 1 = paper
+    elif player_1 == 'paper':
+        if player_2 == 'rock':
+            print("Player 1 wins!")
+        elif:
+            print("Player 2 wins!")
+    # player 1 = scissors
+    elif player_1 == 'scissors':
+        if player_2 == 'rock':
+            print("Player 2 wins!")
+        elif:
+            print("Player 1 wins!")
```
Changed some `print`s to `return`s and fixed some empty `elif`s.
```suggestion
    elif player_1 == 'paper':
        if player_2 == 'rock':
            return "Player 1 wins!"
        else:
            return "Player 2 wins!"
    # player 1 = scissors
    elif player_1 == 'scissors':
        if player_2 == 'rock':
            return "Player 2 wins!"
        else:
            return "Player 1 wins!"
```

alope107 reviewed (commented):
I like it! I would personally be in favor of leaving the answers off and doing it as a classroom activity.

The code that is there is a bit buggy; I think we should probably run & test any code we put in front of students no matter how trivial the function might be. It's super easy to make a little mistake and it really confuses students when the example code doesn't work.

Thanks for putting this all together! This looks like it was a lot of work and I'm excited to see how it goes!

beccaelenzil commented:
Thanks for the review! You're totally right about testing any code we put in front of students, thanks for the reminder. I sent this off for review a little hurriedly.

beccaelenzil commented:
I incorporated your feedback and updated the example to have learn challenges so that it works well as a classroom activity. The code challenge has tests and I confirmed our sample solution passes the tests.

alope107 commented on about-pses/example-pse.md:
```diff
@@ -1,90 +1,372 @@
+Here are some example clarifying questions:
 
-### #4 Steps to Solve a Sub-Problem
+1. How should the function handle invalid user input (i.e. player_1 = ["lizards"](https://bigbangtheory.fandom.com/wiki/Rock,_Paper,_Scissors,_Lizard,_Spock))?
```
I think adding the link is fun but a bit distracting. Would recommend to keep it without and the people who know the reference will still get a laugh :) 

```suggestion
1. How should the function handle invalid user input (i.e. player_1 = "lizards")?
```

alope107 commented on about-pses/example-pse.md:
```diff
@@ -1,90 +1,372 @@
+    result = winner(player_1, player_2)
+
+    # assert
+    assert result == None  
```
```suggestion
    assert result is None  
```

alope107 reviewed (approved):
Looks good to me! Just added two tiny suggestions.

Can you describe how you envision this being taught? That we have them go through it on their own first (with the support of all the answers being present in the explanation boxes) or that they look at it for the first time as we do it as a class?

spitsfire commented on about-pses/example-pse.md:
```diff
@@ -1,90 +1,372 @@
+##### !question
 
-### #2 Two Example Inputs/Outputs
+List five or more questions whose answers would clarify the problem statement
```
What do you all think about lowering this requirement? Personally, I believe that this has forced students in the past to make up questions to ask for the sake of hitting the 5 mark instead of coming up with solid clarifying questions that make sense in the context.

What if we did 3 or more?

beccaelenzil replied:
Agree! Auberon and I had a similar conversation earlier, and I think lowering the requirement might be the right solution!

spitsfire replied:
so maybe: 

```suggestion
List three or more questions whose answers would clarify the problem statement
```
